### ISSUE TITLE
cucumber-expressions: Invalid snippets generated for escaped optional

### ISSUE THREAD

**Issue by mpkorstanje** [cucumber/common#994]
Spotted on stack overflow:

https://stackoverflow.com/questions/61725664/is-it-not-possible-to-have-a-step-definition-with-no-parameters

The following scenario:

```feature
Feature: Spheres
  Scenario: A ray intersects a sphere at two points
    Given r ← ray(point(0, 0, -5), vector(0, 0, 1))
    And s ← sphere()
    When xs ← intersect(s, r)
    Then xs.count = 2
    And xs[0] = 4.0
    And xs[1] = 6.0
```

Trying to keep it simple, I thought I'd write the step definition literally to start off:
```java
import io.cucumber.java8.En

class SphereStepDefinitions: En {
    private val epsilon: Double = 0.00001
    lateinit var s: Sphere
    lateinit var xs : List<Double>

    init {
        Given("s ← sphere\\()") {
            s = Sphere()
        }
    }
}
```
The result,

```
The step "s ← sphere()" is undefined. You can implement it using the snippet(s) below:

Given("s ← sphere\\()", () -> {
    // Write code here that turns the phrase above into concrete actions
    throw new io.cucumber.java8.PendingException();
});
```

The correct step definition is:

```
s ← sphere()
```

Because 

```
    private static final Pattern OPTIONAL_PATTERN = Pattern.compile("(\\\\\\\\)?\\(([^)]+)\\)");
```

Does not allow empty capture groups.

Probably solved by implementing the parser for Cucumber Expressions rather then the current rewriting solution but we should check the snippet generation when we do.

**Comment by lav16kosh** [comment]
@mpkorstanje  I changed the optional_pattern  to (\\\\\\\\)?\\(([^)]*)\\)  to include step definition with no parameters. 
![image](https://user-images.githubusercontent.com/65532932/86214520-ea62fb80-bb98-11ea-9761-b8f91b348966.png)

And I created a test in CucumberExpressionTest.java to verify with no parameters. It worked fine. 

![image](https://user-images.githubusercontent.com/65532932/86214782-3f9f0d00-bb99-11ea-9490-252890138c29.png)

Shall I go ahead and create pull request?

**Comment by mpkorstanje** [comment]
That would work but it would be breaking changes in cucumber expressions. If breaking changes are to be made I'd rather bundle them all up as part of https://github.com/cucumber/cucumber/pull/771

**Comment by aslakhellesoy** [comment]
@mpkorstanje can we close this now? I believe this was fixed as part of #771.
